src/module.c

```diff
@@ -862,6 +893,86 @@ int RM_CreateCommand(RedisModuleCtx *ctx, const char *name, RedisModuleCmdFunc c
     return REDISMODULE_OK;
 }
 
+void extendKeySpecsIfNeeded(struct redisCommand *cmd) {
+    /* We extend even if keys_specs_num == keys_specs_max because
+     * this function is called prior to adding a new spec */
+    if (cmd->keys_specs_num < cmd->keys_specs_max)
+        return;
+
+    cmd->keys_specs_max++;
+
+    if (cmd->keys_specs == cmd->keys_specs_static) {
+        cmd->keys_specs = zmalloc(sizeof(keysSpec) * cmd->keys_specs_max);
+        memcpy(cmd->keys_specs, cmd->keys_specs_static, sizeof(keysSpec) * cmd->keys_specs_num);
+    } else {
+        cmd->keys_specs = zrealloc(cmd->keys_specs, sizeof(keysSpec) * cmd->keys_specs_max);
+    }
+}
+
+int moduleAddCommandKeySpec(RedisModuleCtx *ctx, const char *name, const char *specflags, keysSpec *spec) {
+    int64_t flags = specflags ? commandKeySpecsFlagsFromString((char*)specflags) : 0;
```

Comment:
@guybe7 we need to let modules add arbitrary flags that will be visible to the clients.
let's save the original string in the spec, and reflect that string in the output of COMMAND command.

it would also be nice to do that for the command flags somehow, but maybe as a separate PR.

Reply:
I guess they won't be totally arbitrary, still validated against a set of known flags right?

Reply:
i actually meant totally arbitrary, but that's actually wasn't meant for the key-spec flags, but rather the command flags.
i wanted for modules to add flags such as "all-shards" and such (hinting proxies and clients with some things). but later realized two things:
1. it has nothing to do with this PR (since it's related to the command flags)
2. these flags must be documented in redis.io and in that case they can indeed by validated by redis.
i asked @gkorland and @MeirShpilraien to open another issue to discuss that